dataframe.py

```diff
@@ -0,0 +1,269 @@
+    """
+    A categorical value is an ordinal (integer) value that references a
+    sequence of category values of an arbitrary data type
+    """
+
+    def __init__(self, index_type: IntegerType, category_type: DataType,
+                 ordered: bool = False):
+        self.index_type = index_type
+        self.category_type = category_type
+        self.ordered = ordered
+
+    def equals(self, other: DataType) -> bool:
+        return (isinstance(other, Categorical) and
+                self.index_type == other.index_type and
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    @property
+    def attrs(self) -> Mapping:
+        """
+        Metadata for this column. Default implementation returns empty dict
+        """
+        return {}
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
+        """
+        Return the number of rows in the DataFrame (if known)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def column_names(self) -> Sequence[Any]:
```

Comment:
How about a `columns` property that returns a `ColumnCollection` that itself has a `names` property, and similarly for rows

Reply:
See comment above. I am not sure this API should be in the business of syntactic sugar

Reply:
As a user I'd like to write generic code which would work.regardless of the concrete type of DataFrame passed. For many applications I could probably get by with a restricted API but I'd still want to code a nice interface which used idiomatic python constructs. The proposed interface here strikes me as being extremely cumbersome to use in practice. Just one viewpoint though, there may well be other considerations to take into account other than aesthetics.

Reply:
> As a user I'd like to write generic code which would work.regardless of the concrete type of DataFrame passed.

This is not a requirement / goal of this project (see the original discussion thread).

Reply:
This is *exactly* the motivating use case at the the start of that thread:
> #### Bigger picture: programming to an interface

If the discussion there has veered off from that course then it is probably a good idea to write a more formal proposal to capture the current state.

Reply:
Take it up with people on Discourse. If it turns out that people don't want interchange / export APIs I'm going to walk away from this project. I've already spent a lot of time on it.